Sources/Subprocess/Platforms/Subprocess+Linux.swift

```diff
@@ -265,67 +294,70 @@ extension String {
 // MARK: - Process Monitoring
 @Sendable
 internal func monitorProcessTermination(
-    forExecution execution: Execution
+    for processIdentifier: ProcessIdentifier
 ) async throws -> TerminationStatus {
-    try await withCheckedThrowingContinuation { continuation in
-        _childProcessContinuations.withLock { continuations in
-            // We don't need to worry about a race condition here because waitid()
-            // does not clear the wait/zombie state of the child process. If it sees
-            // the child process has terminated and manages to acquire the lock before
-            // we add this continuation to the dictionary, then it will simply loop
-            // and report the status again.
-            let oldContinuation = continuations.updateValue(continuation, forKey: execution.processIdentifier.value)
-            precondition(oldContinuation == nil)
-
-            // Wake up the waiter thread if it is waiting for more child processes.
-            _ = pthread_cond_signal(_waitThreadNoChildrenCondition)
+    return try await withCheckedThrowingContinuation { continuation in
+        _processMonitorState.withLock { state in
+            switch state {
+            case .notStarted:
+                continuation.resume(throwing: SubprocessError(
+                    code: .init(.failedToMonitorProcess),
+                    underlyingError: nil)
+                )
+            case .failed(let error):
+                continuation.resume(throwing: error)
+            case .started(let storage):
+                // Register processFileDescriptor with epoll
+                var event = epoll_event(
+                    events: EPOLLIN.rawValue,
+                    data: epoll_data(fd: processIdentifier.processFileDescriptor)
+                )
+                let rc = epoll_ctl(
+                    storage.epollFileDescriptor,
+                    EPOLL_CTL_ADD,
+                    processIdentifier.processFileDescriptor,
+                    &event
+                )
+                if rc != 0 {
+                    let error = SubprocessError(
+                        code: .init(.failedToMonitorProcess),
+                        underlyingError: .init(rawValue: errno)
```

Comment:
`errno` is not safe across Swift function calls, so technically you need to save it to a local variable immediately after calling `epoll_ctl` since the initializer you pass to `code` could change it. Worth an audit across the entire codebase as well.